src/task/task.rs

```diff
@@ -0,0 +1,63 @@
+
+use std::time::{Duration, Instant};
+use std::{thread};
+use crate::errors::{ErrorKind, Result};
+
+
+
+/// Status of task
+#[derive(Debug, Clone, Copy)]
+pub enum Status {
+    /// long running task not found
+    NotFound,
+    /// long running task in progress
+    InProgress,
+    /// long running task completed
+    Complete
+}
+
+static POLL_INTERVAL: Duration = Duration::from_secs(5);
+
+
+/// Base task interface
+pub trait Task {
+    /// interface for query specific task status
+    fn query_status(&self) -> Result<Status>;
+
+    /// interface for retrieving client timeout
+    fn get_timeout(&self) -> Result<Duration>;
+
+    // TODO: consider adding async and only support rust 1.39.0+
+    /// logic to wait until status query is completed
+    fn wait_till_complete(&self) -> Result<Status> {
+        let now = Instant::now();
+
+        while now.elapsed().as_secs() < self.get_timeout()?.as_secs() {
```

Comment:
The global client policy timeout applies to commands that are expected to respond immediately, e.g. simple get/put commands, etc. The same timeout should not apply to potentially long-running tasks like creating a new secondary index. If we add a timeout check here, the timeout value should be passed by the app in the `wait_till_complete` function call itself.